marklogic-data-hub-central/src/test/java/com/marklogic/hub/central/controllers/EntitySearchControllerTest.java

```diff
@@ -57,30 +77,122 @@ void testCRUDOnSavedQuery() throws Exception {
+            "                }\n" +
+            "            }\n" +
+            "        },\n" +
+            "        \"propertiesToDisplay\": [\n" +
+            "            \"sku\",\n" +
+            "            \"title\",\n" +
+            "            \"price\",\n" +
+            "            \"discounted-price\",\n" +
+            "            \"order-date\",\n" +
+            "            \"ship-date\"\n" +
+            "        ]\n" +
+            "    }\n" +
+            "}";
+
+        int limit = 5;
+        int headerRow = 1;
+        int totalColumns = 6;
+        int totalRows = limit + headerRow;
+
+        MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
+        params.add("fileType", "csv");
+        params.add("limit", String.valueOf(limit));
+        params.add("queryDocument", json);
+
+        // Export using query document
+        postWithParams(EXPORT_PATH, params)
+            .andExpect(request().asyncStarted())
+            .andDo(MvcResult::getAsyncResult)
+            .andExpect(status().isOk())
+            .andDo(result -> {
+                String response = result.getResponse().getContentAsString();
+                assertEquals(totalRows, response.chars().filter(i -> i == '\n').count() + 1);
+                assertEquals(totalColumns, response.substring(0, response.indexOf("\n")).chars().filter(i -> i == ',').count() + 1);
+            });
+
+        postJson(SAVED_QUERIES_PATH, json)
+            .andExpect(status().isCreated())
+            .andDo(result -> savedQueryResponse = readJsonObject(result.getResponse().getContentAsString()));
+
+        // Export using queryId
+        getJson(EXPORT_PATH + "/query/" + savedQueryResponse.get("savedQuery").get("id").textValue(), params)
+            .andExpect(request().asyncStarted())
+            .andDo(MvcResult::getAsyncResult)
+            .andExpect(status().isOk())
+            .andDo(result -> {
+                String response = result.getResponse().getContentAsString();
+                assertEquals(totalRows, response.chars().filter(i -> i == '\n').count() + 1);
+                assertEquals(totalColumns, response.substring(0, response.indexOf("\n")).chars().filter(i -> i == ',').count() + 1);
+            });
+    }
+
+    private void installExportProject() throws IOException {
+        String projectFolder = "test-projects/export-project";
+        installProjectInFolder(projectFolder);
+        entityManager.deployQueryOptions();
+
+        DocumentMetadataHandle metadata = new DocumentMetadataHandle()
+            .withCollections("Order")
+            .withPermission("data-hub-operator", DocumentMetadataHandle.Capability.READ, DocumentMetadataHandle.Capability.UPDATE);
+        GenericDocumentManager finalDocMgr = getHubConfig().newFinalClient().newDocumentManager();
+        DocumentWriteSet writeSet = finalDocMgr.newWriteSet();
+        File inputDir = new ClassPathResource(projectFolder + "/input").getFile();
```

Comment:
I've worked on projects before where we create a lot of test files that are stored in version control, and it can quickly become difficult to manage. I think it's far better in the long term to write a simple program to generate documents in memory and write them to ML. It's fewer files in version control, and it's easier to change. 

Right now, it doesn't look like the test cares what's in the documents, though I think it should - shouldn't it be verifying that the correct value is present in each column in each row? 

For that reason, I would only insert 2 test documents (generally, there's zero, there's 1, and then there's 2 or more). And I'd only request maybe 3 columns. Now you only need to verify 6 values. And since you have 2 test documents, you can have a separate test that passes in a limit of 1, and you can verify you only get 1 row back. 

With that in mind, I'd design a bare minimum test document - no headers, no triples, no attachments. And if you're only requesting 3 properties, I'd populate 4 total properties - that allows you to verify that when you ask for 3 of them, you don't get the 4th. Then, make a simple method in here that allows you to insert a document with 3 known values. 

Also, let's use the entity-reference-model project. The Customer entity has three atomic properties so far - customerId, name, and customerSince. Feel free to add a fourth atomic property for your test. That avoids having to create a new example project, and the more tests we have using the same sample project, the easier it is to understand the tests.

Reply:
Makes sense. I'll update the tests.